Messages support (in case of assertion failure)

aseemk opened the pull request:
This patch adds an optional `msg` arg to all assertion methods:

```
foo.should.equal(bar, 'baz didn't have an effect!');
all.should.have.length(10, 'elements weren't added')
```

Because this wasn't possible for testing that properties exist:

```
// doesn't work, tests that foo.bar === 'custom message'
foo.should.have.property('bar', 'custom message')
```

I added a new verb called `expose` which is like `property` but doesn't take an optional value:

```
foo.should.expose('bar', 'custom message')
```

Hat tip to @dominictarr for [pointing this out](https://github.com/visionmedia/should.js/issues/1#comment_962613).

This isn't possible for non-method assertions like `ok` and `arguments`, which is why I'd love to change those to methods as well. Haven't figured out a clean way to do that without breaking existing code or having the change be opt-in via a config, which is kinda lame. What are your thoughts on that?

I also improved the static `should.fail()` --- which was just a wrapper for the native `assert.fail()` --- to actually set the passed-in message to the error's `message` property.

Other than that, some minor code and test cleanup. Let me know if something isn't clear.

Cheers,
Aseem

aseemk committed:
Spec'ing should.exist() and should.not.exist().

Readme-driven-development ftw! ;)

aseemk committed:
Test cases for should[.not].exist()!

aseemk committed:
Implementing should[.not].exist()!

aseemk committed:
Adding attribution info.

aseemk committed:
Improved documentation (and semantics) for exist.

aseemk committed:
Updating exist tests to new spec.

aseemk committed:
Test code cleanup.

aseemk committed:
Updated exist implementation.

aseemk committed:
Implementation code cleanup based on review.

aseemk committed:
Merge remote-tracking branch 'visionmedia/master' into feature/exist

* visionmedia/master:
  Release 0.1.0
  cleanup
  Implementation code cleanup based on review.
  Updated exist implementation.
  Test code cleanup.
  Updating exist tests to new spec.
  Improved documentation (and semantics) for exist.
  Adding attribution info.
  Implementing should[.not].exist()!
  Test cases for should[.not].exist()!
  Spec'ing should.exist() and should.not.exist().

aseemk committed:
Slightly better documentation.

aseemk committed:
Moved should.exist tests into main test suite.

aseemk committed:
Fixing readme attribution.

aseemk committed:
Custom assertion messages!

aseemk committed:
Code documentation.

aseemk committed:
New should#expose method, like #property.

aseemk committed:
Fix readme typo.

aseemk committed:
Fixing merge duplication, weird.

aseemk committed:
Merge branch 'feature/exist' into feature/messages

* feature/exist:
  Fixing merge duplication, weird.

aseemk committed:
Improved should.fail().

aseemk commented:
Funny, for some reason it includes all my old commits for `should.exist()` even though you already pulled those in. Any idea why that happens?

aseemk commented on Readme.md:
```diff
@@ -267,7 +303,7 @@ Yes, yes it does, with a single getter _should_, and no it wont break your code,
 (The MIT License)
 
-Copyright (c) 2010-2011 TJ Holowaychuk &lt;tj@vision-media.ca&gt;
+Copyright (c) 2010-2011 TJ Holowaychuk &lt;tj@vision-media.ca&gt;  
```
The two spaces at the end of the line are necessary in Markdown to force a linebreak in the generated HTML. Compare the output in [your repo](https://github.com/visionmedia/should.js#readme) vs [my repo](https://github.com/aseemk/should.js#readme) to see for yourself.

gasi replied:
To avoid having trailing spaces I started using the explicit `<br/>`. Also, for license boilerplates I prefer using block quotes since they originated in source code.

aseemk replied:
By block quotes I assume you meant a `<code>` block, i.e. by indenting four spaces? I agree and that's what I would have done; I was just trying to minimize diff noise for the pull request. ;) Good tip about `<br/>`!

tj commented on Readme.md:
```diff
@@ -204,6 +224,14 @@ Assert property exists and has optional value:
+Note that if you want to assert with a custom message, you _must_ pass a value. If you only want to assert that the property exists, use _expose_:
+
+    // wrong: this will assert that foo.bar === 'custom message'
+    foo.should.have.property('bar', 'custom message')
```
if we keep chaining, why not just have:

```
 foo.should.have.property('bar').message('some random stuff')
```

tj replied:
and defer the assertions, i dunno

aseemk replied:
Well (a) it's much more natural and less verbose to pass the message as an optional param to any assertion you make rather than have a separate message() method, and (b) this change was easy to make whereas I certainly have no idea how one would implement deferred assertions. =P (At least not without some lame flush() method at the end.)

dominictarr replied:
hmm. well the assert could make a new Error object,
  parse the stack trace,
  open the file it was called from,
  look for the line, and then check if the line ends in .property(x)
or .property(x).message('hello')

but that is INSANE.

keys and property are the only assertions which have optional values, so
lastArg is the message would work for everything else.

defuring would be to non-deterministic, the test should fail at the first
failing assertion.

aseemk committed:
Minor code cleanup.

aseemk committed:
Fixing incorrect stack start for static asserts.

aseemk committed:
Minor test code bug fix.

aseemk committed:
Improve static should.not.exist() for errors.

If you pass an error object to should.not.exist() now -- e.g. in the common
Node.js-style callback case -- no custom message is needed. It'll just use
the error itself as the message, e.g. "ReferenceError: foo is not defined".

aseemk committed:
Merge branch 'feature/exist' into feature/messages

* feature/exist:
  Improve static should.not.exist() for errors.
  Minor test code bug fix.
  Fixing incorrect stack start for static asserts.
  Minor code cleanup.

Conflicts:
	lib/should.js (remove stack trace starts)

aseemk commented:
Note that this pull request includes all of the changes in pull request #14. That now includes the automatic error message support:

```
someAsyncTask(function (err, result) {
    should.not.exist(err);    // if this fails, uses err as the message
    // ...
});
```
